### ISSUE TITLE
Add an IRC message handling framework

### ISSUE THREAD

**Issue by aji** [aji/miau#5]
Right now a single `match` is used, which is good enough for the moment. However, a more robust framework for registering IRC command handlers should be implemented.

For the framework to be considered complete, it must meet the following requirements:

  - **A handler can exist in any module.** Due to the limitations of Rust, we cannot do this in an "implied" way, like could be done with decorators in Python or annotations and component scans in Java, but it should still be possible for a closure or function defined in some distant module to be added as a message handler.

  - **Multiple handlers can be registered for the same IRC message, but not given any sort of priority over each other.** The lack of priority ensures predictable message handling behavior.

  - **A handler should be able to alter the `Evented` the message arrived on, and alter the global `Reactor`.** The `Evented` portion of this should seem obvious, but the event loop part is a bit more subtle. We may want to spawn or close a connection while in a message handler, so it's important to be able to alter the `Reactor` as well.

**Comment by BlacklightShining** [comment]
My immediate response to this is to a field to`core::Bot` to store handlers (a `HashMap<Vec<u8>, Vec<Handler>>`?) and provide some kind of interface to let other modules add (and remove?) them. And the last requirement seems pretty easy; `Handler`s could have `Fn`s that take mutable references to the `EventedCtl` and `Reactor`.

**Comment by aji** [comment]
yeah, passing a mutable reference to the `EventedCtl` is possible, but only because `Bot` has both `&mut self` and `&mut EventedCtl` as arguments to `on_line`. the sneaky problem with this (and part of why `EventedCtl` etc. exists) is because when one struct is inside of another, you can't pass mutable references to both without upsetting the borrow checker. that's why `Bot` is not inside of `EventedCtl`, and vice versa.

**Comment by BlacklightShining** [comment]
…hmm, the `Reactor` is local to `Bot::run()`. I think that's gonna have to change if we want to lend it to handlers…could we make it a `Bot` field, too?

**Comment by aji** [comment]
this is blocking all the interesting work, so I'll try to get to it this weekend

**Comment by aji** [comment]
this is sort of handled in the nonsense here now https://github.com/aji/miau/blob/master/src/network.rs#L110-L142

miles ahead of what we had before
